### ISSUE TITLE
Skip button

### ISSUE THREAD

**Issue by zillionn** [MailOnline/videojs-vast-vpaid#56]
Is it possible to always show the skip button for all kinds of ads?

**Comment by Fetz** [comment]
You should only show the skip and execute skip when the ad's says that you can skip, and some ads doesn't support AdSkippableState because that event is new event from VPAID 2.0.

You can add a button instead that will close the ad. But if you do that you will probably lose money

**Comment by indieveed** [comment]
I think it is relevant to ask on this issue:
What is the best way for implementing custom skip button for vast (not vpaid)? There is a custom extension in VAST xml we use with time on which to show skip button. I added custom event and trigger it inside getVastResponse function, so I can get JXON of VAST xml and get from that custom extension the skip time value. But can I reuse skip button code used for VPAID in any way?

**Comment by indieveed** [comment]
Ok, so I tried something like that: 

Change getVastResponse function, treiggering new event: 

```
  function getVastResponse(callback) {
    vast.getVASTResponse(settings.adTagUrl ? settings.adTagUrl() : settings.adTagXML, function(error, response) {
      if (!error) {
        player.trigger({
          type: "vast.adDataLoaded",
          response: response
        });
      }
      if (callback) {
        callback.apply(this, arguments);
      }
    });
  }
```

And then I can change response object, setting new value for skipoffset:

```
          player.on('vast.adDataLoaded', function(response) {
              
              var skipExtension = response.response.ads[0].inLine.extensions.extension.filter(function(el){return el["@type"] == "skipTime2"});
              if (skipExtension.length) {
                var skipTime = skipExtension[0].keyValue.split(":");
                response.response.skipoffset = (skipTime[0] * 60 + skipTime[1]) * 1000;
            }
          });
```

It works, but I am not sure wether it is reliable enough.

**Comment by carpasse** [comment]
@indieveed skippable ads are part of VAST 3.0 and we fully support them for preroll ads.
Please take a look at [VAST 3.0 specification ](http://www.iab.net/media/file/VASTv3.0.pdf) section 2.3.2 Skippable Linear Creative on page 35. 

You can also see how we implemented it [here](https://github.com/MailOnline/videojs-vast-vpaid/blob/master/src/ads/vast/client/VASTIntegrator.js#L168)

Is that not good enough for you?

**Comment by indieveed** [comment]
@carpasse So, the best way for VAST 2.0 ads is to modify original vast response xml and add skip section according to VAST 3.0 spec?

Also, is it possible to get original vast response without modifying the library code? My method with adding custom event seems hackish.

**Comment by carpasse** [comment]
@indieveed So, you would want to use the creative extensions to be able to set the ad as skippable on VAST 2.0 ads. Am I right?

Also, would you just want the original VAST response or also the possibility of modify it?

**Comment by indieveed** [comment]
@carpasse 
Yes, you are right, I have to set VAST 2.0 ad as skippable for a particular ad provider. And also I need to track that event - now I modify the trackingEvents array in original response to achieve that.

I don't think that modifying the original VAST response is a good idea. It would be better if I could make a VAST 2.0 ad skippable without modifying the original response. But some ad providers add their own extensions, which may describe custom trackings, for example. So it is definitely useful to have read access to the original VAST response (mainly for extensions).

But, as I can see, the response returned by getVASTResponse method, is not actually the original XML response, it is parsed in some way. For example, the skipoffset value is parsed and transformed to milliseconds (btw, why to milliseconds).

**Comment by bkolo** [comment]
It would be great to have read access to the original VAST (for error
logging or filtering disapproved ads).  Sometimes there are certain
creatives from marketplaces that publishers just don't want and having
access to the VAST to get the adid would enable blocking those ads.

Is there an existing way to access the  getVASTResponse from
player.on('vast.adStart') or some other way?   I see the
player.vastResponse with type, framework and getSrc but that doesn't quite
have enough info to identify specific creatives.

thanks!

Bob

**Comment by carpasse** [comment]
It seems that a lot of you are looking for a way to intercept calls and to decide when to play the ad.
Give us a some days and we will come with a hooks proposition that will allow you to intercept the different steps that occur while an ad gets played.

**Comment by whatvn** [comment]
@carpasse do you have update for this issue?

**Comment by whatvn** [comment]
I hack a little bit into vast plugin source code to enable skip button with configured skipTime. 
If it can be merged, I would like to make a pull request.

**Comment by TanyuS** [comment]
Please, help! I have video.js v4.12.11 and videojs-vast-vpaid plugin. I need control buttons like "skip ad", "go to site", "close ad". I checked xml - the have such information. For example 
Extension type="skipTime" <![CDATA[00:05]]>
Extension type="skipTime2" <![CDATA[00:05]]>
Extension type="linkTxt" <![CDATA[&#1087;&#1077;&#1088;&#1077;&#1081;&#1090;&#1080; &#1085;&#1072; &#1089;&#1072;&#1081;&#1090; &#1088;&#1077;&#1082;&#1083;&#1072;&#1084;&#1086;&#1076;&#1072;&#1090;&#1077;&#1083;&#1103;]]>
Extension type="isClickable" <![CDATA[1]]>

And also I have such code:
"plugins": {
		'ads-setup': {
			"adsEnabled": true,
			"url": here is url
		}
	}

So ad is shown. But advertiser needs these buttons. I tried to find is this plugin has such options but failed =(

**Comment by vietdien2005** [comment]
just add option skipoffset to your file xml 
Example: 
<Linear skipoffset="00:00:05">
...
 </Linear>
